plugins/microservices/msi_get_open_data_obj_l1desc_index/libmsi_get_open_data_obj_l1desc_index.cpp

```diff
@@ -0,0 +1,83 @@
+/// \file
+
+#include "irods_ms_plugin.hpp"
+#include "irods_re_structs.hpp"
+#include "msParam.h"
+#include "rodsErrorTable.h"
+#include "irods_error.hpp"
+#include "irods_log.hpp"
+#include "objDesc.hpp"
+#include "fileOpr.hpp"
+
+#include <exception>
+#include <functional>
+#include <string>
+
+// persistent L1 object descriptor table
+extern l1desc_t L1desc[NUM_L1_DESC];
+
+namespace
+{
+    int msi_impl(msParam_t* _in, msParam_t* _out, ruleExecInfo_t* _rei)
+    {
+        // check sanity
+        if (!_in || !_out) {
+            return SYS_INTERNAL_NULL_INPUT_ERR;
+        }
+
+        // input param needs to be a string
+        if (strcmp(_in->type, STR_MS_T)) {
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        const char *objPath = (const char*)_in->inOutStruct;
```

Comment:
prefer C++ casts over C casts

Reply:
Sure, generally yes, in this case I believe it would have made no difference since AFAIK no RTTI coming from the other side, and simply casting a void pointer into a char pointer, would compile to the same machine code with `reinterpret_cast<>` and even with `static_cast<>`? Only difference is the syntax, right? (Which I personally do not like to use if don't have to, but of course do use, when casting proper classes etc.).

Fixed, refactored into no cast at all, letting `parseMspForStr` do the type checking, as you've done in all your `msi_` examples. Internally it of course resorts to the C style cast, but later on you can then refactor that function into using what ever C++ cast you wish. This way I also get the included `rodsLog` error reporting.